digitalocean/resource_digitalocean_bucket.go

```diff
@@ -0,0 +1,317 @@
+package digitalocean
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceDigitalOceanBucket() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDigitalOceanBucketCreate,
+		Read:   resourceDigitalOceanBucketRead,
+		Update: resourceDigitalOceanBucketUpdate,
+		Delete: resourceDigitalOceanBucketDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Bucket name",
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Bucket region",
+			},
+			"profile": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    false,
+				Description: "Spaces Access Profile",
+			},
+			"acl": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "Canned ACL applied on bucket creation (Change requires replacement)",
+				Default:     "private",
+			},
+		},
+	}
+}
+
+func resourceDigitalOceanBucketCreate(d *schema.ResourceData, meta interface{}) error {
+	endpoint := fmt.Sprintf("https://%s.digitaloceanspaces.com", d.Get("region").(string))
+	sesh, err := session.NewSession(&aws.Config{
```

Comment:
I would maybe use something like [this](https://github.com/minio/cookbook/blob/master/docs/aws-sdk-for-go-with-minio.md#3-example). The current implementation would require extra work and make it confusing. The proposed change would allow users to also set the access key and secret key via env vars like many other providers and resources.

Reply:
That makes sense.  I committed 064ec7d with what I believe is the correct implementation of this recommendation.

Reply:
Using environment variables is great, but this should also be configurable in the provider block like an access token. This works very much like the schema definition for resources. So `space_access_key` and `spaces_secret_key` can be set as optional and default to the values in the environment variables.

https://github.com/terraform-providers/terraform-provider-digitalocean/blob/master/digitalocean/provider.go#L11

Reply:
Maybe I'm hitting a limit in my golang knowledge but how would one access those values in `provider.go` from my resource?  Since we're dealing with a separate API, I can't seem to get that working.  I was able to get this working by doing pretty much the same thing but defining the parameters directly in my resource, see 9f0f01b.

Reply:
The interface returned from the function set in Provider's `ConfigureFunc` is what gets passed into each resource function as `meta` That is how things configured on that level can be used in a resource. See: https://godoc.org/github.com/hashicorp/terraform/helper/schema#ConfigureFunc

Here's a very quick take on what that might look like: https://gist.github.com/andrewsomething/374a615bd5d6790346290f392ca1ad13

It's a bit more complex, but I think it will make for a better UI in the end.

Reply:
Ah, ok that makes sense.  Thanks, I'll give it a shot!

Reply:
@andrewsomething Thanks for the help with this! I was able to get this working with the AWS Go SDK as represented in 239afd5.  My tests work locally for me but I think you guys will need to generate credentials for the tests running on Github.